### ISSUE TITLE
IStartable components should be started in dependency order

### ISSUE THREAD

**Issue by OZCMRDM** [autofac/Autofac#624]
Service C depends on Service A and B via constructor dependency. All services implement IStartable. In Start() for C, A and B have not yet been started. Logically, IStartable components should be started in dependency order (A,B, then C).

**Comment by jchapuis** [comment]
I have the same issue here, thanks

**Comment by aschaible** [comment]
Same here.  Seems strange that this is not supported.

**Comment by tillig** [comment]
We do accept pull requests... 😉

**Comment by aschaible** [comment]
Yes apologies for the tone - just that Autofac has so much circulation, I was just surprised.  TYVM!

**Comment by tillig** [comment]
No worries. It's just that there are only two project owners for core Autofac as well as all of the supported integrations (I think there are 23 integration and feature libraries now?), so we end up having to prioritize our limited time on the most heavily used / heavily affecting issues. We just ate a little over a full year trying to nail down .NET Core support without much chance to address anything else.

`IStartable` isn't one of the most common things, and `IStartable` _that relies on other `IStartable`_ makes it even more edge case. Combine that with the fact that the solution isn't super quick to implement and this sort of thing gets prioritized down the list a bit.

From a design perspective, you sort of have to walk the entire resolution chain for every `IStartable` and build the dependency chain; then determine if any one `IStartable` depends on any other `IStartable`; then resolve/execute accordingly, which _again_ walks that dependency tree, this time actually doing the resolution/instantiation.

It becomes a challenge when you start factoring in additional things like lambda registrations - you can resolve things using `context.Resolve<T>()` inside a lambda registration; what if one of those uses or provides an `IStartable` (or both)? You don't really know what happens inside that lambda until you execute it.

Sooooo... there's a non-trivial amount of thought that has to go into implementing the feature and making sure it works in all the crazy edge cases in which folks are expecting it to "just work," and given the other issues that are pending out there, we just haven't gotten to this one.

**We truly would be open to a pull request to help with this.** If you're interested, drop some thoughts on the design in here so we can talk about it **before you spend a lot of time coding**. We should make sure all the use cases are thought through and can be addressed before we can take code to solve the issue.

**Comment by ggrnd0** [comment]
Why `IStartable` needed if `IRegistrationBuilder.OnActivated, etc...` methods exist and provide high control of component lifetime?

How this problems solved for `IRegistrationBuilder.OnActivated, etc...`? Is them solved?

**Comment by tillig** [comment]
I think this is going to be one of those "known issues" things where you can solve it by using `SingleInstance` with `OnActivated` as needed instead of `IStartable` if you run into challenges. With the latest Autofac, you can even use `RegisterBuildCallback` to have things run on container build.

Contrived example:

```csharp
public class TestClass
{
  private class Dependency1
  {
    public Dependency1(ITestOutputHelper output)
    {
      output.WriteLine("Dependency1.ctor");
    }
  }

  private class Dependency2
  {
    private ITestOutputHelper output;

    public Dependency2(ITestOutputHelper output, Dependency1 dependency)
    {
      this.output = output;
      output.WriteLine("Dependency2.ctor");
    }

    public void Initialize()
    {
      this.output.WriteLine("Dependency2.Initialize");
    }
  }

  private class Dependency3
  {
    private ITestOutputHelper output;

    public Dependency3(ITestOutputHelper output, Dependency1 dependency)
    {
      this.output = output;
      output.WriteLine("Dependency3.ctor");
    }

    public void Initialize()
    {
      this.output.WriteLine("Dependency3.Initialize");
    }
  }

  private class Dependency4
  {
    private ITestOutputHelper output;

    public Dependency4(ITestOutputHelper output, Dependency2 dependency2, Dependency3 dependency3)
    {
      this.output = output;
      output.WriteLine("Dependency4.ctor");
    }

    public void Initialize()
    {
      this.output.WriteLine("Dependency4.Initialize");
    }
  }

  // Xunit passes this to the ctor of the test class 
  private ITestOutputHelper _output;

  public TestClass(ITestOutputHelper output)
  {
    this._output = output;
  }

  [Fact]
  public void OnActivatedDependencyChain()
  {
    var builder = new ContainerBuilder();
    builder.RegisterInstance(this._output).As<ITestOutputHelper>();
    builder.RegisterType<Dependency1>().SingleInstance();

    // The OnActivated replaces the need for IStartable. When an instance
    // is activated/created, it'll run the Initialize method as specified. Using
    // SingleInstance means that only happens once.
    builder.RegisterType<Dependency2>().SingleInstance().OnActivated(args => args.Instance.Initialize());
    builder.RegisterType<Dependency3>().SingleInstance().OnActivated(args => args.Instance.Initialize());
    builder.RegisterType<Dependency4>().SingleInstance().OnActivated(args => args.Instance.Initialize());

    // Notice not really in dependency order - 4 needs 2 and 3 to complete starting.
    builder.RegisterBuildCallback(c => c.Resolve<Dependency4>());
    builder.RegisterBuildCallback(c => c.Resolve<Dependency2>());
    builder.RegisterBuildCallback(c => c.Resolve<Dependency1>());
    builder.RegisterBuildCallback(c => c.Resolve<Dependency3>());

    // This will run the build callbacks.
    var container = builder.Build();

    // These effectively do NOTHING. OnActivated won't be called again
    // because they're SingleInstance.
    container.Resolve<Dependency1>();
    container.Resolve<Dependency2>();
    container.Resolve<Dependency3>();
    container.Resolve<Dependency4>();
  }
}
```

The test output you get here is:

```
Dependency1.ctor
Dependency2.ctor
Dependency3.ctor
Dependency4.ctor
Dependency2.Initialize
Dependency3.Initialize
Dependency4.Initialize
```

Note it's all in dependency order as expected.

**Comment by tillig** [comment]
[Docs updated with detailed explanation and example.](http://docs.autofac.org/en/latest/lifetime/startup.html)

**Comment by tillig** [comment]
Fix released in [v4.7.0](https://www.nuget.org/packages/Autofac/4.7.0).
